Source/DafnyStandardLibraries/src/Std/TargetSpecific/Duration.dfy

```diff
@@ -0,0 +1,405 @@
+include "DateTimeConstant.dfy"
+include "../Collections/SeqExt.dfy"
+
+module Std.Duration {
+  import opened DateTimeConstant
+  import opened Strings
+  import opened Collections.SeqExt
+  import opened Collections.Seq
+  import opened BoundedInts
+  import opened Arithmetic.Power
+
+  datatype Duration = Duration(
+    seconds: uint16,
+    millis: uint16
+  ) {
+
+    ghost predicate Valid() {
+      millis < (MILLISECONDS_PER_SECOND as uint16)
+    }
+  }
+
+  function ToTotalMilliseconds(d: Duration): uint64
+    requires d.Valid()
+  {
+    var product := (d.seconds as uint64) * (MILLISECONDS_PER_SECOND as uint64);
+    var sum := product + (d.millis as uint64);
+    sum
+  }
+
+  function FromMilliseconds(ms: uint32): Duration
+    ensures FromMilliseconds(ms).Valid()
+  {
+    /**
+    Duration((ms / (MILLISECONDS_PER_SECOND as uint32)) as uint16,
+             (ms % (MILLISECONDS_PER_SECOND as uint32)) as uint16)*/
+    var ms64 := ms as uint64;
+    var secondsValue := ms64 / (MILLISECONDS_PER_SECOND as uint64);
+    assume  {:axiom} secondsValue < 65536;
```

Comment:
Be very careful using `assume` in production code. You have to be really confident the statement is actually always true, but have no practical way of proving it. In this case it's not actually true in general, which means for large `uint32` values you could have overflow, which is exactly the kind of runtime error we want to use Dafny to avoid.

If you take my suggestion to use a `uint64` for seconds in Duration, I don't think you'll need any of these assumptions.